src/kernel/chainparamsbase.cpp

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <kernel/chainparamsbase.h>
+
+#include <tinyformat.h>
+
+const std::string CBaseChainParams::MAIN = "main";
+const std::string CBaseChainParams::TESTNET = "test";
+const std::string CBaseChainParams::SIGNET = "signet";
+const std::string CBaseChainParams::REGTEST = "regtest";
+
+/**
+ * Port numbers for incoming Tor connections (8334, 18334, 38334, 18445) have
+ * been chosen arbitrarily to keep ranges of used ports tight.
+ */
+std::unique_ptr<CBaseChainParams> CreateBaseChainParams(const std::string& chain)
+{
+    if (chain == CBaseChainParams::MAIN) {
+        return std::make_unique<CBaseChainParams>("", 8332, 8334);
+    } else if (chain == CBaseChainParams::TESTNET) {
+        return std::make_unique<CBaseChainParams>("testnet3", 18332, 18334);
+    } else if (chain == CBaseChainParams::SIGNET) {
+        return std::make_unique<CBaseChainParams>("signet", 38332, 38334);
+    } else if (chain == CBaseChainParams::REGTEST) {
+        return std::make_unique<CBaseChainParams>("regtest", 18443, 18445);
+    }
+    throw std::runtime_error(strprintf("%s: Unknown chain %s.", __func__, chain));
+}
+
+static std::unique_ptr<CBaseChainParams> globalChainBaseParams;
```

Comment:
Hmmm... Been a while since I last took a look but how hard would it be to have this as a member of `kernel::Context`? What does the caller graph look like?

It might be the case that this is a situation where it'd be too intrusive to have the N I C E C O D E as discussed here: https://github.com/bitcoin/bitcoin/pull/25065#pullrequestreview-985450940

Reply:
The caller graph is kind of annoying, it boils down to users of `SelectBaseParams` (like bitcoin-cli), and more importantly `SelectParams`, which is used all over the place (I count 23 separate files as call site). If I understand the current usage of `kernel::Context` correctly, this would mean that we'd have to construct the context for nearly each call site separately, and many in places where it was previously not required. I currently tend towards handling this in a separate refactor attempt to get rid of these param globals in general.

Reply:
You could maybe split it up? Have chainparams provide `SelectParams(ArgsManager&)` and use that instead of `SelectParams(args.GetChainName())` which covers all the cases outside of test/ and test/fuzz/. Then most of the test or test/fuzz cases already hardcode the network, and could be replaced by `kernel::SelectParams(RegTestBase(), RegTest(regtestopts))` that just initializes the globals?

Reply:
I've had no luck coming up with a nice pattern to get this split done without blowing everything up :/
I might try it again at a later point in time by first create an ugly solution and then see if I can mold it into something nice.

Reply:
I looked into this again. The underlying problem seems to be that the kernel library still depends on the `ArgsManager`, which in turn depends on the global. I'd rather not make this change devolve into changing `ArgsManager` internals and behaviour, so I'll defer refactoring this global to once the `ArgsManager` is removed from the kernel library.